In [1]:
#  Ebnable HTML/CSS 
from IPython.core.display import HTML
HTML("<link href='https://fonts.googleapis.com/css?family=Passion+One' rel='stylesheet' type='text/css'><style>div.attn { font-family: 'Helvetica Neue'; font-size: 30px; line-height: 40px; color: #FFFFFF; text-align: center; margin: 30px 0; border-width: 10px 0; border-style: solid; border-color: #5AAAAA; padding: 30px 0; background-color: #DDDDFF; }hr { border: 0; background-color: #ffffff; border-top: 1px solid black; }hr.major { border-top: 10px solid #5AAA5A; }hr.minor { border: none; background-color: #ffffff; border-top: 5px dotted #CC3333; }div.bubble { width: 65%; padding: 20px; background: #DDDDDD; border-radius: 15px; margin: 0 auto; font-style: italic; color: #f00; }em { color: #AAA; }div.c1{visibility:hidden;margin:0;height:0;}div.note{color:red;}</style>")

___
Enter Team Member Names here (double click to edit):

- Name 1: Nikhil Gupta
- Name 2: Hayley Horn
- Name 3: Max Moro
- Name 4: Michael Tieu


________

# In Class Assignment Three
In the following assignment you will be asked to fill in python code and derivations for a number of different problems. Please read all instructions carefully and turn in the rendered notebook (or HTML of the rendered notebook)  before the end of class.

<a id="top"></a>
## Contents
* <a href="#Loading">Loading the Data</a>
* <a href="#distance">Measuring Distances</a>
* <a href="#KNN">K-Nearest Neighbors</a>
* <a href="#naive">Naive Bayes</a>

________________________________________________________________________________________________________
<a id="Loading"></a>
<a href="#top">Back to Top</a>
## Downloading the Document Data
Please run the following code to read in the "20 newsgroups" dataset from sklearn's data loading module.

In [2]:
from sklearn.datasets import fetch_20newsgroups_vectorized
import numpy as np
from __future__ import print_function

# this takes about 30 seconds to compute, read the next section while this downloads
ds = fetch_20newsgroups_vectorized(subset='train')

# this holds the continuous feature data (which is tfidf)
print('features shape:', ds.data.shape) # there are ~11000 instances and ~130k features per instance
print('target shape:', ds.target.shape) 
print('range of target:', np.min(ds.target),np.max(ds.target))
print('Data type is', type(ds.data), float(ds.data.nnz)/(ds.data.shape[0]*ds.data.shape[1])*100, '% of the data is non-zero')

features shape: (11314, 130107)
target shape: (11314,)
range of target: 0 19
Data type is <class 'scipy.sparse.csr.csr_matrix'> 0.1214353154362896 % of the data is non-zero


## Understanding the Dataset
Look at the description for the 20 newsgroups dataset at http://qwone.com/~jason/20Newsgroups/. You have just downloaded the "vectorized" version of the dataset, which means all the words inside the articles have gone through a transformation that binned them into 130 thousand features related to the words in them.  

**Question Set 1**:
- How many instances are in the dataset? 
- What does each instance represent? 
- How many classes are in the dataset and what does each class represent?
- Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not?
- Is the data represented as a sparse or dense matrix?

___
Enter your answer here:

- We have 11,314 Instances. But these are the 'train' side (60%). The 'test' side should have 7,542 istance (total is 18,856)
- Each Instance represent an article
- We have 130,107 classes, each class represent a word
- I may expect the classifier would work for documents written in the past weeks and from same newsgroup, This is based on the assumption  the articles from same newsgoup may contain similar set of words, but this could NOT be true if the articles are about very specific and different topics or different language.
- I think this would be a sparse matrix as only a small subset of the 130K are used in the article hence many values in the matrix are zeros.





___
<a id="distance"></a>
<a href="#top">Back to Top</a>
## Measures of Distance
In the following block of code, we isolate three instances from the dataset. The instance "`a`" is from the group *computer graphics*, "`b`" is from from the group *recreation autos*, and "`c`" is from group *recreation motorcycle*. **Exercise for part 2**: Calculate the: 
- (1) Euclidean distance
- (2) Cosine distance 
- (3) Jaccard similarity 

between each pair of instances using the imported functions below. Remember that the Jaccard similarity is only for binary valued vectors, so convert vectors to binary using a threshold. 

**Question for part 2**: Which distance seems more appropriate to use for this data? **Why**?

In [3]:
from scipy.spatial.distance import cosine
from scipy.spatial.distance import euclidean
from scipy.spatial.distance import jaccard
import numpy as np

# get first instance (comp)
idx = 550
a = ds.data[idx].todense()
a_class = ds.target_names[ds.target[idx]]
print('Instance A is from class', a_class)

# get second instance (autos)
idx = 4000
b = ds.data[idx].todense()
b_class = ds.target_names[ds.target[idx]]
print('Instance B is from class', b_class)

# get third instance (motorcycle)
idx = 7000
c = ds.data[idx].todense()
c_class = ds.target_names[ds.target[idx]]
print('Instance C is from class', c_class)

# Enter distance comparison below for each pair of vectors:
p = 'Placeholder'
print('\n\nEuclidean Distance\n ab:', euclidean(a, b), 'ac:', euclidean(a, c), 'bc:', euclidean(b, c))
print('Cosine Distance\n ab:', cosine(a, b), 'ac:', cosine(a, c), 'bc:', cosine(b, c))
print('Jaccard Dissimilarity (vectors should be boolean values)\n ab:', jaccard(a > 0, b> 0), 'ac:', jaccard(a> 0, c> 0), 'bc:', jaccard(b> 0, c> 0))

print("\nThe most appropriate distance is cosine. It delineates topics effectively.")

Instance A is from class comp.graphics
Instance B is from class rec.autos
Instance C is from class rec.motorcycles


Euclidean Distance
 ab: 1.0985184671870858 ac: 1.1891405425398234 bc: 0.9177794226661624
Cosine Distance
 ab: 0.6033714113755322 ac: 0.7070276149559529 bc: 0.4211595343347173
Jaccard Dissimilarity (vectors should be boolean values)
 ab: 0.8821138211382114 ac: 0.8754716981132076 bc: 0.9087947882736156

The most appropriate distance is cosine. It delineates topics effectively.


___
# Start of Live Session Assignment
___
<a id="KNN"></a>
<a href="#top">Back to Top</a>
## Using scikit-learn with KNN
Now let's use stratified cross validation with a holdout set to train a KNN model in `scikit-learn`. Use the example below to train a KNN classifier. The documentation for `KNeighborsClassifier` is here: http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html  

**Exercise for part 3**: Use the code below to test what value of `n_neighbors` works best for the given data. *Note: do NOT change the metric to be anything other than `'euclidean'`. Other distance functions are not optimized for the amount of data we are working with.* 

**Question for part 3**: What is the accuracy of the best classifier you can create for this data (by changing only the `n_neighbors` parameter)? 

In [4]:
from sklearn.cross_validation import StratifiedShuffleSplit
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from IPython.html import widgets 

cv = StratifiedShuffleSplit(ds.target, n_iter = 1, test_size = 0.5, train_size=0.5)

# fill in the training and testing data and save as separate variables
for trainidx, testidx in cv:
    # note that these are sparse matrices
    X_train = ds.data[trainidx] 
    X_test = ds.data[testidx] 
    y_train = ds.target[trainidx]
    y_test = ds.target[testidx]

# fill in your code  here to train and test
# calculate the accuracy and print it for various values of K


best_K = 0
best_acc = 0
for i in range(1,11):
    K = i 
    clf = KNeighborsClassifier(n_neighbors = K, weights = 'uniform', metric = 'euclidean')

    knn_model = clf.fit(X_train, y_train)
    acc = clf.score(X_test, y_test)
    
    if acc > best_K:
        best_acc = acc
        best_K = i
    print('Accuracy of classifier with %d neighbors is: %.2f'%(K, acc))
#=====================================

print('\nThe best accuracy is %.2f with %d neighbors'%(best_acc, best_K))


/Users/mtieu/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/Users/mtieu/anaconda3/lib/python3.6/site-packages/IPython/html.py:14: ShimWarning: The `IPython.html` package has been deprecated since IPython 4.0. You should import from `notebook` instead. `IPython.html.widgets` has moved to `ipywidgets`.
  "`IPython.html.widgets` has moved to `ipywidgets`.", ShimWarning)


Accuracy of classifier with 1 neighbors is: 0.59
Accuracy of classifier with 2 neighbors is: 0.55
Accuracy of classifier with 3 neighbors is: 0.53
Accuracy of classifier with 4 neighbors is: 0.51
Accuracy of classifier with 5 neighbors is: 0.50
Accuracy of classifier with 6 neighbors is: 0.49
Accuracy of classifier with 7 neighbors is: 0.48
Accuracy of classifier with 8 neighbors is: 0.47
Accuracy of classifier with 9 neighbors is: 0.46
Accuracy of classifier with 10 neighbors is: 0.45

The best accuracy is 0.59 with 1 neighbors


**Question for part 3**: With sparse data, does the use of a KDTree representation make sense? Why or Why not?

Enter your answer below:

With sparse data, the use of KDTree representation does not make sense. Due to the high number of "zero" value attributes, the resulting tree will need to be very deep. The cost to calculate distances is very low for sparse data, and the cost for creating a deep tree is high. Hence, using a KDTree will not be beneficial for a sparse matrix. 

_____
## KNN extensions - Centroids
Now lets look at a very closely related classifier to KNN, called nearest centroid. In this classifier (which is more appropriate for big data scenarios and sparse data), the training step is used to calculate the centroids for each class. These centroids are saved. Unknown attributes, at prediction time, only need to have distances calculated for each saved centroid, drastically decreasing the time required for a prediction. 

**Exercise for part 4**: Use the template code below to create a nearest centroid classifier. Test which metric has the best cross validated performance: Euclidean, Cosine, or Manhattan. In `scikit-learn` you can see the documentation for NearestCentroid here: 
- http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestCentroid.html#sklearn.neighbors.NearestCentroid

and for supported distance metrics here:
- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.distance_metrics.html#sklearn.metrics.pairwise.distance_metrics

In [5]:
from sklearn.neighbors.nearest_centroid import NearestCentroid

# the parameters for the nearest centroid metric to test are:
#    l1, l2, and cosine (all are optimized)

clf_list = ["l2", "cosine", "l1"]

# fill in your code here
best_acc = 0
best_distance = ""
for distance in clf_list:
    clf = NearestCentroid(metric = distance)
    
    model = clf.fit(X_train, y_train)    
    acc = clf.score(X_test, y_test)
    
    if acc > best_acc:
        best_acc = acc
        best_distance = distance
    print(distance, acc)
    
print("\nThe best distance metric is: ", best_distance, " with an accuracy of ", best_acc)

l2 0.41841965706204703


/Users/mtieu/anaconda3/lib/python3.6/site-packages/sklearn/neighbors/nearest_centroid.py:140: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


cosine 0.4864769312356373
l1 0.32384656178186316

The best distance metric is:  cosine  with an accuracy of  0.4864769312356373


___
<a id="naive"></a>
<a href="#top">Back to Top</a>
## Naive Bayes Classification
Now let's look at the use of the Naive Bayes classifier. The 20 newsgroups dataset has 20 classes and about 130,000 features per instance. Recall that the Naive Bayes classifer calculates a posterior distribution for each possible class. Each posterior distribution is a multiplication of many conditional distributions: 

$${\arg \max}_{j} \left(p(class=j)\prod_{i} p(attribute=i|class=j) \right)$$

where $p(class=j)$ is the prior and $p(attribute=i|class=j)$ is the conditional probability.

**Question for part 5**: With this many classes and features, how many different conditional probabilities need to be parameterized? How many priors need to be parameterized?

In [6]:
# Use this space for any calculations you might want to do
ds.data.shape[1] * len(set(ds.target))

2602140

Enter your answer here:

Number of Conditional Probabilities = 2,602,140 (130,107 * 20)<br>
Number of Priors = Number of Classes = 20

___
## Naive Bayes in Scikit-learn
Scikit has several implementations of the Naive Bayes classifier: `GaussianNB`, `MultinomialNB`, and `BernoulliNB`. Look at the documentation here: http://scikit-learn.org/stable/modules/naive_bayes.html Take a look at each implementation and then answer this question: 

**Questions for part 6**: 
- If the instances contain mostly continuous attributes, would it be better to use Gaussian Naive Bayes, Multinomial Naive Bayes, or Bernoulli? And Why?

- What if the data is sparse, does this change your answer? Why or Why not?


Enter you answer here:

The Gaussian Naive Bayes would be the best option to use for continuous attributes (with the assumption of a normal distribution of the instances). Bernoulli is not a good use here since it is better for binary attributes rather than continuous ones. Multinomial Naive Bayes is an extension of the Bernoulli with k > 2, but is used for classification with discreet features.


If the data were sparse, the Gaussian Naive Bayes would not be the best option to use since a gaussian distribution will not capture the data accurately (most of the data will be 0). For sparse data, it is probably better to discretize the data into two or more classes and use Multinomial Naive Bayes or Bernoulli.
___

## Naive Bayes Comparison
For the final section of this notebook let's compare the performance of Naive Bayes for document classification. Look at the parameters for `MultinomialNB`, and `BernoulliNB` (especially `alpha` and `binarize`). 

**Exercise for part 7**: Using the example code below, change the parameters for each classifier and see how accurate you can make the classifiers on the test set. 

**Question for part 7**: Why are these implementations so fast to train? What does the `'alpha'` value control in these models (*i.e.*, how does it change the parameterizations)? 

In [7]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB




# fill in your code here
a = 0.01


increments = -0.001

mnb_acc = 0
mnb_alpha = 0.0

bnb_acc = 0
bnb_alpha = 0.0
bnb_binarize = 0.0

while a > 0.0001:
    b = 0.01
    clf_mnb = MultinomialNB(alpha = a)

    clf_mnb.fit(X_train, y_train) 
    
    model1_acc = clf_mnb.score(X_test, y_test)
    
    if model1_acc > mnb_acc:
        mnb_acc = model1_acc
        mnb_alpha = a
    #print("mnb acc:", mnb_acc)
        
    while b > 0.0001:
        clf_bnb = BernoulliNB(alpha = a, binarize = b)
        clf_bnb.fit(X_train, y_train)
        
        model2_acc = clf_bnb.score(X_test, y_test)
        
        if model2_acc > bnb_acc:
            bnb_acc = model2_acc
            bnb_binarize = b
            bnb_alpha = a
        #print("bnb acc:", bnb_acc)
        b += increments
    a += increments

print("Best accuracy for MNB: ", mnb_acc, ". Alpha:", mnb_alpha)
print("Best accuracy for BNB: ", bnb_acc, ". Alpha:", bnb_alpha, "Binarize:", bnb_binarize)

print("\nThese classifiers are so fast because the algorithms only need to calculate probabilities in the training step. There is no optimization step involved.")
print("\nThe alpha values control the smoothing factor, which is useful when the conditional probabilities go to zero. This changes paramaterization by adding a constant term to the numerator and denominator account for any missing data.")


Best accuracy for MNB:  0.8849213363973838 . Alpha: 0.001
Best accuracy for BNB:  0.8684815273112957 . Alpha: 0.001 Binarize: 0.01

These classifiers are so fast because the algorithms only need to calculate probabilities in the training step. There is no optimization step involved.

The alpha values control the smoothing factor, which is useful when the conditional probabilities go to zero. This changes paramaterization by adding a constant term to the numerator and denominator account for any missing data.


________________________________________________________________________________________________________

That's all! Please **upload your rendered notebook to blackboard** and please include **team member names** in the notebook submission.